src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+
+## Matching
+
+A pattern is always tested against a value to determine if the value has the form
+you expect, or in other words, if the value _matches_ the pattern. 
+
+What constitutes a match depends on the [pattern type][types].
+For example, matching a constant pattern means the value is equivalent to the pattern:
+
+```dart
+switch (number) {
+  case 1: ...        // Constant pattern 1 matches if number == 1
+}
+```
+
+Many patterns make use of subpatterns (sometimes called outer and inner patterns,
+respectively). Patterns match recursively on their subpatterns.
+For example, the individual fields of any [collection-type][] pattern could be 
+[variable patterns][variable] or [constant patterns][constant]:
+
+```dart
+switch (obj) {       // List pattern [a,b] matches obj first if obj
+  case [a, b]: ...   // is a list wth two fields, then if its  
+}                    // fields match the constant subpatterns a and b  
+```
+
+Patterns can also _partially_ match, by using a [wildcard pattern][] as a place holder,
+or a [rest element][] in the case of list patterns.
+
+If you need to ensure `null` values aren't silently treated as match failures,
+use the [null-check pattern][] while matching.
+
+## Destructuring
+
+When an object and pattern match, the pattern can then access the object's data 
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+var [a, b, c] = numList;   // List pattern [a, b, c] destructures the list object
+print(a + b + c);          // numList into individual new parts. Prints 6.
+```
+
+You can use the [null-check pattern][] to check for `null` while destructuring. 
+
+All patterns attempt to match the objects they're compared against, but not all
+patterns destructure.
+Whether a pattern destructures depends on the [type of pattern][types]
+and the [context](#pattern-uses) in which it's being used.
+
+## Pattern uses
+
+Patterns integrate into several code constructs across the Dart language:
+
+- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
+- [for and for-in loops][for]
+- [if-cases] and [switch-cases]
+- [Collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns, 
+combining various contexts, pattern types, and surrounding code constructs.
+
+### Variable declaration
+
+Anywhere Dart allows regular, single-name local variable declaration, you can also
+use a _pattern variable declaration_. If the pattern matches the value on the right
+of the declaration, it destructures the value, binding it to new local variables:
+
+```dart
+var (a, [b, c]) = ("str", [1, 2]);   // Declares new variables a, b, and c.
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+Use the [null-assert pattern][] to eliminate `null` in variable declarations.
```

Comment:
This should probably be phrased more carefully.  It eliminates `null` by throwing an exception, which makes this dangerous as blanket advice.

Reply:
+1. I just wouldn't talk about `!` here.